src/sst/elements/vanadis/inst/vfpmadd.h

```diff
@@ -0,0 +1,142 @@
+            ((sizeof(fp_format) == 8) && (VANADIS_REGISTER_MODE_FP32 == isa_opts->getFPRegisterMode())) ? 6 : 3,
+            ((sizeof(fp_format) == 8) && (VANADIS_REGISTER_MODE_FP32 == isa_opts->getFPRegisterMode())) ? 3 : 1)
+    {
+        if ( (sizeof(fp_format) == 8) && (VANADIS_REGISTER_MODE_FP32 == isa_opts->getFPRegisterMode()) ) {
+            isa_fp_regs_in[0]  = src_1;
+            isa_fp_regs_in[1]  = src_1 + 1;
+            isa_fp_regs_in[2]  = src_2;
+            isa_fp_regs_in[3]  = src_2 + 1;
+            isa_fp_regs_in[4]  = src_3;
+            isa_fp_regs_in[5]  = src_3 + 1;
+            isa_fp_regs_out[0] = dest;
+            isa_fp_regs_out[1] = dest + 1;
+        }
+        else {
+            isa_fp_regs_in[0]  = src_1;
+            isa_fp_regs_in[1]  = src_2;
+            isa_fp_regs_in[2]  = src_3;
+            isa_fp_regs_out[0] = dest;
+        }
+    }
+
+    VanadisFPFusedMultiplyInstruction*  clone() override { return new VanadisFPFusedMultiplyInstruction(*this); }
+    VanadisFunctionalUnitType getInstFuncType() const override { return INST_FP_ARITH; }
+
+    const char* getInstCode() const override
+    {
+        if ( std::is_same<fp_format, double>::value ) {
+			return "FP64MADD";
+		  }
+        else if ( std::is_same<fp_format, float>::value ) {
+            return "FP32MADD";
+        }
+        else {
+            return "FPMADDUNK";
+        }
+    }
+
+    void printToBuffer(char* buffer, size_t buffer_size) override
+    {
+        snprintf(
+            buffer, buffer_size,
+            "%s  %5" PRIu16 " <- %5" PRIu16 " * %5" PRIu16 " + %5" PRIu16 " (phys: %5" PRIu16 " <- %5" PRIu16 " * %5" PRIu16 " + %5" PRIu16 ")",
+            getInstCode(), isa_fp_regs_out[0], isa_fp_regs_in[0], isa_fp_regs_in[1], isa_fp_regs_in[2], phys_fp_regs_out[0],
+            phys_fp_regs_in[0], phys_fp_regs_in[1], phys_fp_regs_in[2]);
+    }
+
+    void execute(SST::Output* output, VanadisRegisterFile* regFile) override
+    {
+#ifdef VANADIS_BUILD_DEBUG
+        if ( output->getVerboseLevel() >= 16 ) {
+            output->verbose(
+                CALL_INFO, 16, 0, "Execute: 0x%llx %s\n", getInstructionAddress(), getInstCode());
+        }
+#endif
+
+        if ( (sizeof(fp_format) == 8) && (VANADIS_REGISTER_MODE_FP32 == isa_options->getFPRegisterMode()) ) {
+            const fp_format src_1  = combineFromRegisters<fp_format>(regFile, phys_fp_regs_in[0], phys_fp_regs_in[1]);
+            const fp_format src_2  = combineFromRegisters<fp_format>(regFile, phys_fp_regs_in[2], phys_fp_regs_in[3]);
+            const fp_format src_3  = combineFromRegisters<fp_format>(regFile, phys_fp_regs_in[4], phys_fp_regs_in[5]);
+            const fp_format result = std::fma( (fp_format) src_1, (fp_format) src_2, (fp_format) src_3 ); 
+
+            if ( std::fetestexcept(FE_INVALID) ) {
+                fpflags.setInvalidOp();
+                update_fp_flags = true;
+            }
```

Comment:
Are you doing this to trap exceptions generated by the host? Because you should clear the exception flags before to make sure an exception didn't occur earlier in execution.

Reply:
I think it is better to perform the checks against values than look at the exception flags on the host.